olm/operator-bundle.md

```diff
@@ -0,0 +1,199 @@
+* The label `operators.coreos.com.bundle.resources` represents the bundle type:
+    * The value `manifests` implies that this bundle contains operator manifests only.
+    * The value `metadada` implies that this bundle has operator metadata only.
+    * The value `manifests+metadata` implies that this bundle contains both operator metadata and manifets.
+* The label `operators.coreos.com.bundle.mediatype` reflects the media type of the operator bundle. It could be helm charts, plain kubernetes manifests etc.
+
+The labels will also be put inside a YAML file, as shown below.
+*annotations.yaml*
+```yaml
+annotations:
+  operators.coreos.com.bundle.resources: "manifests+metadata"
+  operators.coreos.com.bundle.mediatype: bundle
+```
+
+The package manifest YAML file that defines a set of `channel`(s) will not be included in the bundle. Operator registry will provide machineries to define `package` and `channel` that can be used while index is being built. 
+
+For an operator registry bundle, the source directory structure will have similar layout as shown below.
+```
+$ tree test
+test
+├── 0.1.0
+│   ├── testbackup.crd.yaml
+│   ├── testcluster.crd.yaml
+│   ├── testoperator.v0.1.0.clusterserviceversion.yaml
+│   └── testrestore.crd.yaml
+└── annotations.yaml
+```
+
+`Dockerfile` for operator bundle
+```
+FROM scratch
+
+# We are pushing an operator-registry bundle that
+# has both metadata and manifests.
+LABEL operators.coreos.com.bundle.resources=manifests+metadata
+LABEL operators.coreos.com.bundle.mediatype=bundle
+
+ADD test/0.1.0 /manifests
+ADD test/annotations.yaml /metadata/annotations.yaml
+```
+
+Below is the directory layout of the operator bundle insdie the image.
+```bash
+$ tree
+/
+├── manifests
+│   ├── testbackup.crd.yaml
+│   ├── testcluster.crd.yaml
+│   ├── testoperator.v0.1.0.clusterserviceversion.yaml
+│   └── testrestore.crd.yaml
+└── metadata
+    └── annotations.yaml
+```
+
+###### UX:
+Build, Push and Pull an operator bundle.
+```
+docker build -f Dockerfile -t quay.io/test/test-operator:v1 .
+docker push quay.io/test/test-operator:v1
+docker pull quay.io/test/test-operator:v1
+```
+
+A tool can inspect an operator bundle to determine the bundle type and its media.
+```bash
+docker image inspect quay.io/test/test-operator:v1 | \
+jq '.[0].Config.Labels["operators.coreos.com.bundle.resources"]'
+
+"manifests+metadata"
+```
+
+### Verify, Run and Test 
+
+#### Generate Scaffolding
+As an operator author I want to generate the scaffolding necessary to create an operator bundle. We provide the operator author to automatically generate the scaffolding.
+```bash
+$ tree test
+test
+├── 0.1.0
+│   ├── testbackup.crd.yaml
+│   ├── testcluster.crd.yaml
+│   ├── testoperator.v0.1.0.clusterserviceversion.yaml
+│   └── testrestore.crd.yaml
+
+$ cd test
+
+# the following command generates the necessary scaffolding.
+$ operator-framework bundle init --type=operator-registry --bundle-folder=0.1.0
```

Comment:
If we merge this proposal are we committing to the `operator-framework` cli name?

Reply:
No, it's just an example. :) It's an open question whether we have a new cli, or use `operator-sdk` or `operator-registry`.